pkg/cmd/roachtest/tests/mixed_version_c2c.go

```diff
@@ -0,0 +1,135 @@
+		workload:                  replicateKV{readPercent: 0, debugRunDuration: 1 * time.Minute, maxBlockBytes: 1, initWithSplitAndScatter: true},
+		additionalDuration:        0 * time.Minute,
+		cutover:                   30 * time.Second,
+		skipNodeDistributionCheck: true,
+		suites:                    registry.Suites(registry.Nightly),
+	}
+
+	r.Add(registry.TestSpec{
+		Name:             "c2c/mixed-versions",
+		Owner:            registry.OwnerDisasterRecovery,
+		Cluster:          r.MakeClusterSpec(sp.dstNodes + sp.srcNodes + 1),
+		CompatibleClouds: sp.clouds,
+		Suites:           registry.Suites(registry.Nightly),
+		Run: func(ctx context.Context, t test.Test, c cluster.Cluster) {
+			runC2CMixedVersions(ctx, t, c, sp)
+		},
+	})
+}
+
+// TODO (msbutler): schedule upgrades during initial scan and cutover.
+func runC2CMixedVersions(ctx context.Context, t test.Test, c cluster.Cluster, sp replicationSpec) {
+
+	sourceMvt := mixedversion.NewTest(ctx, t, t.L(), c, c.Range(1, sp.srcNodes),
+		mixedversion.MinimumSupportedVersion("v23.2.0"),
+		mixedversion.AlwaysUseLatestPredecessors)
+
+	destMvt := mixedversion.NewTest(ctx, t, t.L(), c, c.Range(sp.srcNodes+1, sp.srcNodes+sp.dstNodes),
+		mixedversion.MinimumSupportedVersion("v23.2.0"),
+		mixedversion.AlwaysUseLatestPredecessors,
+	)
+
+	rd := makeReplicationDriver(t, c, sp)
+	cleanup := rd.setupC2C(ctx, t, c)
+	defer cleanup()
+
+	var pgUrl chan *url.URL
+
+	sourceTenantName := "source"
+	destTenantName := "destination"
+
+	// Start Replication
+	sourceMvt.OnStartup("start source", func(ctx context.Context, l *logger.Logger, r *rand.Rand, h *mixedversion.Helper) error {
+		srcNode := c.Node(1)
+		srcClusterSetting := install.MakeClusterSettings()
+		addr, err := c.ExternalPGUrl(ctx, t.L(), srcNode, roachprod.PGURLOptions{})
+		if err != nil {
+			return err
+		}
+		pgURL, err := copyPGCertsAndMakeURL(ctx, t, c, srcNode, srcClusterSetting.PGUrlCertsDir, addr[0])
+		if err != nil {
+			return err
+		}
+		pgUrl <- pgURL
+		return nil
+	})
+	destMvt.OnStartup("start replication", func(ctx context.Context, l *logger.Logger, r *rand.Rand, h *mixedversion.Helper) error {
+		sourcePGURL := <-pgUrl
+		h.System.Exec(r, fmt.Sprintf("CREATE TENANT %q FROM REPLICATION OF %q ON '%s'", destTenantName, sourceTenantName, sourcePGURL.String()))
+		return nil
+	})
+
+	// Start Workload
+	var tpccInitCmd, tpccRunCmd *roachtestutil.Command
+	stopWorkload := sourceMvt.Workload("tpcc", c.Range(1, sp.srcNodes), tpccInitCmd, tpccRunCmd)
+
+	var destHasFinalized chan struct{}
+
+	// @renato: i dont think this should be called several times
+	sourceMvt.InMixedVersion("wait for dest to finalize", func(ctx context.Context, l *logger.Logger, r *rand.Rand, h *mixedversion.Helper) error {
+		<-destHasFinalized
+		return nil
+	})
+
+	destMvt.AfterUpgradeFinalized("cutover and allow source to finalize", func(ctx context.Context, l *logger.Logger, r *rand.Rand, h *mixedversion.Helper) error {
```

Comment:
i'm assuming `AfterUpgradeFinalized` is run exactly once at the end of `mvt.Run()`

Reply:
It's called at the end of each upgrade. So if the test plan generated is from 24.1 -> 24.2 -> master, it will be called two times (once the cluster finalizes the upgrade to 24.2, and once the cluster finalizes the upgrade to master).

If you are only interested in the upgrade to `master` (or release branch where test is running), you can use `h.Context().ToVersion.Equal(clusterupgrade.CurrentVersion())`.